03-data-containers-and-repetitions-3.md

```diff
@@ -0,0 +1,420 @@
+
+        return write_to(self.data, fname)
+
+n = Norse('./norse.json')
+n.to_json('./norse_processed.json')  # output to ./norse_processed.json
+```
+
+Of course, the content of the `norse_processed.json` file should be the same as the input `norse.json` file, as we have yet to make any changes.
+
+### With context manager
+
+The "`with`" statement involves an interesting Python mechanism known as runtime context management; you can read more details about it on its [official documentation](https://docs.python.org/3.8/reference/compound_stmts.html#with). In this particular case, the object returned by the built-in `open()` function implements such a context manager, eliminating the need for some chores such as closing the underlying file system I/O connectivity when it is no longer needed (or when an exception occurs).
+
+Without the context manager, our `read_from()` function implementation may look like so:
+
+```python
+def read_from(fname):
+    try:
+        f = open(fname, mode='r')
+        return json.load(f)
+    except:
+        raise
+    finally:
+        try:
+            f.close()
+        except:
+            pass
+```
+
+The "`finally`" keyword is an exception control on top of the basic `try/except` we have seen already. The logic within its scope happens _regardless of_ and _after_ whether an exception occurs or not. On top of that, we issue a lazily suppressed inner `try/except` logic around the `f.close()` method invocation. Still, in reality, there may be more intricacies that need careful handling.
+
+With a context manager, none of the above is necessary for the users to manage.
+
+### Classes or Functions?
+
+On many occasions, encapsulating data and expose access and operations to them through object properties and methods offer no distinct edge over simple functions. There is no settled convention of class usages among the Python or, to a certain extent, all applicable programming languages communities. As demonstrated, one viable approach is to derive out generic algorithms (such as `read_from()` and `write_to()`) in functions outside of classes while keeping only _stateful_ ones (dependent on the current property value of `data`, such as `to_json()`) inside.
+
+Let us take the `flatten_norse()` implementation from [the previous exercise problem](03-data-containers-and-repetitions-2.md#exercises), apply it to mutate the internal state (`data`) to have each of its rows flattened. The built-in `enumerate()` function is used to generate an indexed iterator, so we have easy access to both the list's index and the corresponding individual item:
+
+```python
+'''norse_type.py'''
+# ...
+
+def flatten_norse(row):
+    flat = {}
+
+    for k, v in row.items():
+        if type(v) is not dict:
+            flat[k] = v
+        else:
+            for nk, nv in v.items():
+                flat['{0}.{1}'.format(k, nk)] = nv
+
+    return flat
+
+def flatten_func(data):  # function equiv of flatthen() method
```

Comment:
very small typo
```suggestion
def flatten_func(data):  # function equiv of flatten() method
```

Reply:
lol there were 2 more types that actually were more significant :trollface: thanks!